cmds/modprobe/modprobe.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2013-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// modprobe - Add and remove modules from the Linux Kernel
+//
+// Synopsis:
+//     modprobe modulename...
+//
+// Author:
+//     Roland Kammerer <dev.rck@gmail.com>
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/u-root/u-root/pkg/kmodule"
+	"golang.org/x/sys/unix"
+)
+
+const cmd = "modprobe modulename"
+
+type modState uint8
+
+const (
+	_ modState = iota
+	loading
+	loaded
+)
+
+type dependency struct {
+	state modState
+	deps  []string
+}
+
+type depMap map[string]*dependency
+
+var moduleDir = "/lib/modules/" // gets extended in main
+
+func genDeps(rd io.Reader) (depMap, error) {
+	deps := make(depMap)
+
+	scanner := bufio.NewScanner(rd)
+	for scanner.Scan() {
+		txt := scanner.Text()
+		nameDeps := strings.Split(txt, ":")
+		modPath, modDeps := nameDeps[0], nameDeps[1]
+		modPath = strings.TrimSpace(modPath)
+
+		var dependency dependency
+		if len(modDeps) > 0 {
+			for _, dep := range strings.Split(strings.TrimSpace(modDeps), " ") {
+				dependency.deps = append(dependency.deps, dep)
+			}
+		}
+		deps[modPath] = &dependency
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return deps, nil
+}
+
+func findModPath(modName string, m depMap) (string, error) {
+	for mp := range m {
+		if strings.HasSuffix(mp, modName+".ko") {
+			return mp, nil
+		}
+	}
+
+	return "", fmt.Errorf("modprobe: Could not find path for module %s", modName)
+}
+
+func loadModule(modPath, options string) error {
+	filename := filepath.Join(moduleDir, modPath)
+	f, err := os.Open(filename)
+	if err != nil {
+		return fmt.Errorf("modprobe: Could not open %q: %v", filename, err)
+	}
+	defer f.Close()
+
+	if err, e := kmodule.FileInit(f, options, 0); err != nil && e != unix.EEXIST {
+		return fmt.Errorf("modprobe: could not load %q: %v", filename, err)
+	}
+
+	return nil
+}
+
+func loadDeps(modPath string, m depMap) error {
+	dependency, ok := m[modPath]
+	if !ok {
+		return fmt.Errorf("modprobe: Could not find dependency %s", modPath)
+	}
+
+	if dependency.state == loading {
+		return fmt.Errorf("modprobe: Circular dependency, %s already in state LOADING", modPath)
+	} else if dependency.state == loaded {
+		return nil
+	}
+
+	m[modPath].state = loading
+
+	for _, dep := range dependency.deps {
+		if err := loadDeps(dep, m); err != nil {
```

Comment:
You can make this faster by checking if the dependency is loaded. If a dependency is loaded, you do not have to attempt loading the dependency's dependencies (and the dependencies of those dependencies, etc...). The whole sub-tree can be cut off.

The height of the dependency graph is fairly small, so this optimization might not be worth applying.

Reply:
Maybe I overlook something, but there is already a check for "== loaded". If a we have the same dependencies in two subtrees, the second one is cut anyways.

Reply:
in main ~159 it first tries to load dependencies before the "top module" is loaded. There we could first try to load the top module and check EEXISTS before even starting with the dependencies. Yes, that could make sense.

Reply:
Sorry, I wasn't specific enough. I meant that the kernel can be queried to see if the module has already been loaded. If a module is already loaded in the kernel, it should be true that all the dependencies are also loaded.

Probably the easiest method is to query the kernel via `/proc/modules` and for each module it contains, initialize the `state` values in `depMap` to `loaded`. Do this before calling `loadDeps`.

Reply:
Hm, I'm not a fan of that approach, I deliberately decided against that in the first place. I mean we can do all kinds of fancy querying (like parsing `/proc/modules`). Eventually we have to decide that we try to load *now*. Things can change in between, unlikely, but possible. So we have to handle all these cases like the module is already loaded anyways. Why not use the actual syscall and its return as the source of truth, which it is, and instead try to rely on outdated information we parsed back then? The worst case is a handfull syscalls that return EEXISTS.

Reply:
I would have to agree with @rck here. I would do the /proc/modules thing only if modprobe turns out to be too slow.

Reply:
Yeah, this optimization is not worth applying.